source/utils/security.py

```diff
@@ -177,19 +200,111 @@ def isObjectAboveLockScreen(obj: "NVDAObjects.NVDAObject") -> bool:
 	):
 		return True
 
-	runningAppModules = appModuleHandler.runningTable.values()
-	lockAppModule = next(filter(_isLockAppAndAlive, runningAppModules), None)
-
+	from appModuleHandler import _checkWindowsForAppModules
+	lockAppModule = _findLockAppModule()
 	if lockAppModule is None:
-		# lockAppModule not running/registered by NVDA yet
-		log.debugWarning(
-			"lockAppModule not detected when Windows is locked. "
-			"Cannot detect if object is in lock app, considering object as insecure. "
+		_checkWindowsForAppModules()
+		lockAppModule = _findLockAppModule()
+		if lockAppModule is None:
+			# lockAppModule not running/registered by NVDA yet
+			log.debug(
+				"lockAppModule not detected when Windows is locked. "
+				"Cannot detect if object is above lock app, considering object as safe. "
+			)
+			return True
+
+	from NVDAObjects.window import Window
+	if not isinstance(obj, Window):
+		log.debug(
+			"Cannot detect if object is above lock app, considering object as safe. "
 		)
-	elif lockAppModule is not None and obj.processID == lockAppModule.processID:
+		# Must be a window instance to get the HWNDVal, other NVDAObjects do not support this.
 		return True
 
-	return False
+	topLevelWindowHandle = winUser.getAncestor(obj.windowHandle, winUser.GA_ROOT)
+	return _isObjectAboveLockScreenCheckZOrder(topLevelWindowHandle, lockAppModule.processID)
+
+
+def _isObjectAboveLockScreenCheckZOrder(objWindowHandle: int, lockAppModuleProcessId: int) -> bool:
+	"""
+	This is a risky hack.
+	If the order is incorrectly detected,
+	secure information may become accessible.
+
+	If these functions fail, where possible,
+	NVDA should make NVDA objects accessible.
+	"""
+
+	def _isWindowLockApp(hwnd: winUser.HWNDVal) -> bool:
+		windowProcessId, _threadId = winUser.getWindowThreadProcessID(hwnd)
+		return windowProcessId == lockAppModuleProcessId
+
+	try:
+		return _isWindowAboveWindowMatchesCond(objWindowHandle, _isWindowLockApp)
+	except _UnexpectedWindowCountError:
+		log.debugWarning("Couldn't find lock screen")
+		return True
+
+
+class _UnexpectedWindowCountError(Exception):
+	"""
+	Raised when a window which matches the expected condition
+	is not found by _isWindowAboveWindowMatchesCond
+	"""
+	pass
+
+
+def _isWindowAboveWindowMatchesCond(
+		window: winUser.HWNDVal,
+		matchCond: Callable[[winUser.HWNDVal], bool]
+) -> bool:
+	"""
+	This is a risky hack.
+	The order may be incorrectly detected.
+
+	@returns: True if window is above a window that matches matchCond.
+	If the first window is not found, but the second window is,
+	it is assumed that the first window is above the second window.
+
+	In the context of _isObjectAboveLockScreenCheckZOrder, NVDA starts at the lowest window,
+	and searches up towards the closest/lowest lock screen window.
+	If the lock screen window is found before the NVDAobject, the object should be made accessible.
+	This is because if the NVDAObject is not present, we want to make it accessible by default.

```

Comment:
I think this sentence is helpful to include still. amending to be more clear.

```suggestion
	This is because if the NVDAObject is not present we want to make it accessible by default.
```